cylc/flow/network/graphql.py

```diff
@@ -0,0 +1,372 @@
+
+    def has_arg_val(self, arg_name, arg_value):
+        """Search through document definitions for argument value.
+
+        Args:
+            arg_name (str): Field argument to search for.
+            arg_value (Any): Argument value required.
+
+        Returns:
+
+            Boolean
+
+        """
+        for components in self.operation_defs.values():
+            defn = components['definition']
+            if (
+                    defn.operation not in STRIP_OPS
+                    or getattr(
+                        defn.name, 'value', None) == 'IntrospectionQuery'
+            ):
+                continue
+            if self.args_selection_search(
+                    components['definition'].selection_set,
+                    components['variables'],
+                    components['parent_type'],
+                    arg_name,
+                    arg_value,
+            ):
+                return True
+        return False
+
+    def args_selection_search(
+            self, selection_set, variables, parent_type, arg_name, arg_value):
+        """Recursively search through feild/fragment selection set fields."""
+        for field in selection_set.selections:
+            if isinstance(field, ast.FragmentSpread):
+                if field.name.value in self.visited_fragments:
+                    continue
+                frag_def = self.fragment_defs[field.name.value]
+                frag_type = type_from_ast(self.schema, frag_def.type_condition)
+                if self.args_selection_search(
+                        frag_def.selection_set, variables,
+                        frag_type, arg_name, arg_value):
+                    return True
+                self.visited_fragments.add(frag_def.name)
+                continue
+            field_def = get_field_def(
+                self.schema, parent_type, field.name.value)
+            if field_def is None:
+                continue
+            arg_vals = get_argument_values(
+                field_def.args, field.arguments, variables)
+            if arg_vals.get(arg_name) == arg_value:
+                return True
+            if field.selection_set is None:
+                continue
+            if self.args_selection_search(
+                    field.selection_set, variables,
+                    get_named_type(field_def.type), arg_name, arg_value):
+                return True
+        return False
+
+
+def execute_and_validate(
+        schema,  # type: GraphQLSchema
+        document_ast,  # type: Document
+        *args,  # type: Any
+        **kwargs  # type: Any
+):
+    # type: (...) -> Union[ExecutionResult, Observable]
+    """Validate schema, and execute request doc against it."""
+    do_validation = kwargs.get("validate", True)
+    if do_validation:
+        validation_errors = validate(schema, document_ast)
+        if validation_errors:
+            return ExecutionResult(errors=validation_errors, invalid=True)
```

Comment:
This code is identical to the original can we wrap it to avoid duplication.

Reply:
Done.

Reply:
Thanks, just need to slot this dependency into the `setup.py` as it is only installed ATM due to being a dependency of another cylc dependency.

Reply:
No point with this PR...
- Graphene has it as `"graphql-core>=2.1,<3",`
- Anything fundamental and less than 3 isn't going to change (as they are on beta release of 3).
- We will address this when we upgrade to graphql-core 3+.